### PR TITLE
Add pbm backup action

### PR DESCRIPTION
**Pull request by MiaAltieri**
## Summary
adding pbm backup action

## Context
Backup requires starting the snap `pbm.agent` agent service. Starting pbm agent requires a reachable pbm URI. The pbm URI is not reachable until it is created. 

The pbm user is created by the leader in `on_start` meaning that the URI cannot be set on until the leader unit has created the user. But it isn't gauranteed that `create-backup` action will be run on the leader. To gaurentee that the agent is running on the unit that `create-backup` is being called on it is best to start `pbm.agent` in the `backup-action` code. Note starting the `pbm` agent is idempotent.

Note that it is necessary to wait for `pbm.agent` to be fully started hence the `sleep`

## Manual test
Manually tested by `juju ssh`-ing into the mongodb unit and setting up configurations:
```
percona-backup-mongodb  config --set storage.type="s3"
percona-backup-mongodb  config --set storage.s3.region="us-west-2"
percona-backup-mongodb  config --set storage.s3.bucket="pbm-test-bucket-1"
percona-backup-mongodb  config --set storage.s3.prefix="data/pbm/backup"
percona-backup-mongodb  config --set storage.s3.credentials.access-key-id="*"
percona-backup-mongodb  config --set storage.s3.credentials.secret-access-key="*"
percona-backup-mongodb  config --set storage.s3.serverSideEncryption.sseAlgorithm="aws:kms"
percona-backup-mongodb  config --set storage.s3.serverSideEncryption.kmsKeyID="*"
```
then exiting and running `juju run-action mongodb/0 create-backup`

## Future PRs
As mentioned this is the first PR of several to get the backup action up and running. The rest are:
- adding config options with s3 integrator
- Add unit tests
- Integration tests

### CODE REVIEW

File: config.yaml [hunk 1]
```diff
@@ -5,6 +5,36 @@ options:
   auto-delete:
     type: boolean
     description: |
-        When a relation is removed, auto-delete ensures that any relevant databases
-        associated with the relation are also removed
+      When a relation is removed, auto-delete ensures that any relevant databases
+      associated with the relation are also removed
     default: false
+
```
Comment (delgod) [inline config.yaml:11]: I think it shouldn't change `config.yaml` here

File: lib/charms/mongodb/v0/mongodb.py [hunk 2]
```diff
@@ -321,7 +321,7 @@ def create_role(self, role_name: str, privileges: dict, roles: dict = []):
         except OperationFailure as e:
             if not e.code == 51002:  # Role already exists
                 logger.error("Cannot add role. error=%r", e)
-                raise e
+                raise
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb.py:324]: changed in another PRm will be merge conflict

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 3]
```diff
@@ -46,27 +47,32 @@ def __init__(self, charm, substrate="k8s"):
...
+
+    def _on_create_backup_action(self, event) -> None:
+        try:
+            snap_cache = snap.SnapCache()
+            pbm_snap = snap_cache["percona-backup-mongodb"]
+            if not pbm_snap.present:
+                logger.debug("Cannot start PBM agent, PBM snap is not yet installed.")
+                event.defer()
+                return
+            pbm_snap.start(services=["pbm-agent"])
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:60]: just Q: is it possible to configure storage before `pbm-agent` is started?

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 4]
```diff
@@ -46,27 +47,32 @@ def __init__(self, charm, substrate="k8s"):
...
-
-    def _on_pbm_config_changed(self, event) -> None:
-        """Handles PBM configurations."""
-        # handling PBM configurations requires that the pbm snap is installed.
-        if "db_initialised" not in self.charm.app_peer_data:
-            logger.debug("Cannot set PBM configurations, MongoDB has not yet started.")
-            event.defer()
+        self.framework.observe(self.charm.on.create_backup_action, self._on_create_backup_action)
+
+    def _on_create_backup_action(self, event) -> None:
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:52]: I think we should check if the s3 relation was created before running backup?

**Review by delgod — CHANGES_REQUESTED** [review]

File: config.yaml (hunk 1 above)
Comment (MiaAltieri) [reply config.yaml:11]: thanks for spotting this was an articifact from `git checkout -b`

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 4 above)
Comment (MiaAltieri) [reply lib/charms/mongodb/v0/mongodb_backups.py:52]: yes, great point thank you Mykola

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 3 above)
Comment (MiaAltieri) [reply lib/charms/mongodb/v0/mongodb_backups.py:60]: Yes, it is :)

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 5]
```diff
@@ -46,26 +48,35 @@ def __init__(self, charm, substrate="k8s"):
...
+            time.sleep(10)
+        except snap.SnapError as e:
+            logger.error("An exception occurred when starting pbm agent, error: %s.", str(e))
+            event.fail(f"Failed to backup MongoDB with error: {str(e)}")
             return
 
-        # TODO handle PBM configurations set by the user
-        self.charm.unit.status = ActiveStatus()
+        try:
+            subprocess.check_output("percona-backup-mongodb backup", shell=True)
```
Comment (marceloneppel) [inline lib/charms/mongodb/v0/mongodb_backups.py:75]: Don't you need to specify the `PBM_MONGODB_URI`  environment variable or the `--mongodb-uri` command line argument when running the backup?

**Review by marceloneppel — COMMENTED** [review]
Nice work, Mia! There is a fix that can be made on `metadata.yaml` to avoid errors in the charm.

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 6]
```diff
@@ -36,6 +37,7 @@
     ("storage.s3.credentials.secret-access-key", "s3-secret-access-key"),
     ("storage.s3.serverSideEncryption.kmsKeyID", "s3-kms-key-id"),
 ]
+S3_RELATION = "s3-credentials"
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply lib/charms/mongodb/v0/mongodb_backups.py:40]: good to know!  It is defined in my s3-relation PR :)

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 5 above)
Comment (MiaAltieri) [reply lib/charms/mongodb/v0/mongodb_backups.py:75]: This is set when the s3-relation is joined (seperate PR)

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 7]
```diff
@@ -46,26 +48,35 @@ def __init__(self, charm, substrate="k8s"):
...
+        except snap.SnapError as e:
+            logger.error("An exception occurred when starting pbm agent, error: %s.", str(e))
+            event.fail(f"Failed to backup MongoDB with error: {str(e)}")
             return
 
-        # TODO handle PBM configurations set by the user
-        self.charm.unit.status = ActiveStatus()
+        try:
+            subprocess.check_output("percona-backup-mongodb backup", shell=True)
+            event.set_results({"backup-status": "backup started"})
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply lib/charms/mongodb/v0/mongodb_backups.py:76]: great idea, thank you Marcelo!

**Review by marceloneppel — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Pull request merged**